spring-pulsar/src/test/java/org/springframework/pulsar/listener/DefaultPulsarMessageListenerContainerTests.java

```diff
@@ -202,4 +204,60 @@ void negativeAckRedeliveryBackoff() throws Exception {
 		pulsarClient.close();
 	}
 
+	@Test
+	void deadLetterPolicy() throws Exception {
+		Map<String, Object> config = new HashMap<>();
+		config.put("topicNames", Collections.singleton("dpmlct-016"));
+		config.put("subscriptionName", "dpmlct-sb-016");
+		config.put("ackTimeoutMillis", 1);
+		DeadLetterPolicy deadLetterPolicy = DeadLetterPolicy.builder().maxRedeliverCount(1).deadLetterTopic("dlq-topic")
+				.build();
+		config.put("deadLetterPolicy", deadLetterPolicy);
+
+		final PulsarClient pulsarClient = PulsarClient.builder().serviceUrl(getPulsarBrokerUrl()).build();
+		final DefaultPulsarConsumerFactory<String> pulsarConsumerFactory = new DefaultPulsarConsumerFactory<>(
+				pulsarClient, config);
+
+		CountDownLatch dlqLatch = new CountDownLatch(1);
+		CountDownLatch latch = new CountDownLatch(6);
+
+		PulsarContainerProperties dlqContainerProperties = new PulsarContainerProperties();
+		dlqContainerProperties
+				.setMessageListener((PulsarRecordMessageListener<?>) (consumer, msg) -> dlqLatch.countDown());
+		dlqContainerProperties.setSchema(Schema.STRING);
+		dlqContainerProperties.setSubscriptionType(SubscriptionType.Shared);
+		dlqContainerProperties.setTopics(new String[] { "dlq-topic" });
+		DefaultPulsarMessageListenerContainer<String> dlqContainer = new DefaultPulsarMessageListenerContainer<>(
+				pulsarConsumerFactory, dlqContainerProperties);
+		dlqContainer.start();
+
+		PulsarContainerProperties pulsarContainerProperties = new PulsarContainerProperties();
+		pulsarContainerProperties.setMessageListener((PulsarRecordMessageListener<?>) (consumer, msg) -> {
+			latch.countDown();
+			if (((String) msg.getValue()).endsWith("4")) {
```

Comment:
Another alternative would be to check the latch current count instead of the message payload.

Reply:
I refactored it now to use an Integer message type and moved to one-based indexing, which improves readability quite a bit